### PR TITLE
mpi: add helper function to print PID and hostname and wait

### PR DESCRIPTION
**Pull request by davydden**
I was looking around how to attach a debugger without running
```
mpirun -np 4 xterm -e lldb -f <my_exec> -- <params>
```
or alike.

I think this could be useful for users and developers.

### CODE REVIEW

**Comment by masterleinad** [comment]
I am not too much in favor of putting this into the library since this technique is not immediately related to the library and this code fragment modifies the normal execution in a very specific way (also running in `Release` mode is affected).

On the other hand, I see value in mentioning this when we teach to use `MPI` in combination with `deal.II`, i.e. `step-40`.

**Comment by davydden** [comment]
@masterleinad 

> I am not too much in favor of putting this into the library

I am fine if it does not go in, just saying if something goes wrong, you could quickly use it to debug a broken `deal.II` unit test or a broken user library. That was an intention.

File: include/deal.II/base/mpi.h [hunk 1]
```diff
@@ -85,6 +85,14 @@ namespace Utilities
      */
     unsigned int this_mpi_process (const MPI_Comm &mpi_communicator);
 
+    /**
+     * This function will print process id and host name on each MPI process
+     * and wait untill the user enters any character before proceeding.
```
Comment (bangerth) [inline include/deal.II/base/mpi.h]: until

File: include/deal.II/base/mpi.h [hunk 2]
```diff
@@ -85,6 +85,14 @@ namespace Utilities
...
+     * It is intendent to facilitate attachment of serial debuggers to MPI programs,
```
Comment (bangerth) [inline include/deal.II/base/mpi.h]: intended

File: include/deal.II/base/mpi.templates.h [hunk 3]
```diff
@@ -378,6 +378,28 @@ namespace Utilities
     {
       internal::all_reduce(MPI_MIN, values, mpi_communicator, minima);
     }
+
+
+
+    void wait(const MPI_Comm &mpi_communicator)
```
Comment (bangerth) [inline include/deal.II/base/mpi.templates.h]: the function is not a template -- it should be in the .cc file

File: include/deal.II/base/mpi.templates.h [hunk 4]
```diff
@@ -378,6 +378,28 @@ namespace Utilities
...
+    {
+      // see https://www.open-mpi.org/faq/?category=debugging#serial-debuggers
+      char hostname[256];
+      ::gethostname(hostname, sizeof(hostname));
```
Comment (bangerth) [inline include/deal.II/base/mpi.templates.h]: We have a function for this in namespace `Utilities`.

**Comment by davydden** [comment]
@dealii/dealii 

> I am not too much in favor of putting this into the library

so let's make it easy and formal via a poll :

👍   in favour
👎   against
😕   no opinion

**Comment by davydden** [comment]
p.s. hopefully, i will add usage example to this thread soon...

File: source/base/mpi.cc [hunk 5]
```diff
@@ -61,6 +61,23 @@ namespace Utilities
...
+
+      char a;
+      if (Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
+        {
+          std::cout << "Type any character and press Enter/Return to start: " << std::flush;
+          std::cin >> a;
+        }
+
+#ifdef DEAL_II_WITH_MPI
+      MPI_Bcast(&a, 1, MPI_BYTE, 0, mpi_communicator);
```
Comment (Rombur) [inline source/base/mpi.cc]: Why are usin `MPI_Bcast`? Shouldn't you use `MPI_Barrier`?

Comment (davydden) [reply source/base/mpi.cc]: agree, barrier should be enough.

Comment (bangerth) [reply source/base/mpi.cc]: It doesn't actually make a difference. In a barrier, everyone has to wait till everyone else gets here. With the bcast, all but processor 0 have to wait, but processor 0 can continue as soon as it sends its data to all others, who are then free to continue once they receive the signal.

**Comment by davydden** [comment]
@dealii/dealii ok, here's proof of the pudding:

Say I have a buggy `deal.II` test which fails on 2 cores 
```
diff --git a/tests/mpi/collective_tensor.cc b/tests/mpi/collective_tensor.cc
index 7f9266e511..b6b81d3816 100644
--- a/tests/mpi/collective_tensor.cc
+++ b/tests/mpi/collective_tensor.cc
@@ -28,6 +28,12 @@ void test()
   unsigned int myid = Utilities::MPI::this_mpi_process (MPI_COMM_WORLD);
   const unsigned int numprocs = Utilities::MPI::n_mpi_processes (MPI_COMM_WORLD);
 
+  if (myid==1)
+    {
+      int i = 10;
+      AssertThrow(i*5==1, ExcInternalError());
+    }
+
   if (myid==0)
     deallog << "Running on " << numprocs << " CPU(s)." << std::endl;
 
@@ -95,6 +101,7 @@ int main(int argc, char *argv[])
 {
 #ifdef DEAL_II_WITH_MPI
   Utilities::MPI::MPI_InitFinalize mpi_initialization (argc, argv, testing_max_num_threads());
+  Utilities::MPI::wait(MPI_COMM_WORLD);
 #else
   (void)argc;
   (void)argv;
```

(1) I add the new function and run the test
```
mpiexec -np 2 xterm -e tests/mpi/collective_tensor.debug/collective_tensor.debug
```
screen shot 2018-03-21 at 22 16 17

(2) then I know PID and ranks, so in this case I can use any IDE GUI which can attach to PID. For example, in Visual Studio Code I can add _attach_ debug launch configuration with
```
    {
      "name": "(lldb) Attach",
      "type": "cppdbg",
      "request": "attach",
      "program": "${workspaceFolder}/_build/tests/mpi/collective_tensor.debug/collective_tensor.debug",
      "processId": "${command:pickProcess}",
      "MIMode": "lldb"
    },
```
IDE is clever enough to allow me choose the process ID when I run it thanks to `"processId": "${command:pickProcess}",` setting. And then, everything works as you would expect from IDE+serial debugger:

screen shot 2018-03-21 at 22 18 07

Bottom line: if you don't have cash for TotalView, but want something a bit more convenient than running all debugger command-lines manually (`bt`, etc), then this is IMO a good compromise. 
I would expect the same approach to work in other IDE so long as they can attach to the given process ID.
